### DISCUSSION TITLE
Sidekiq slowdown

### DISCUSSION THREAD

**Discussion by thatcosmonaut** [mastodon/mastodon#17861] [Q&A]
Been running an instance for a few years now. For the past week, about once per day, the default Sidekiq queue gets extremely backlogged and latency of up to an hour occurs. This wasn't happening until fairly recently. I've been digging through logs and performance monitoring but I can't find anything obvious. Any idea why this is happening?

**Comment by ClearlyClaire** [comment]
Nothing immediately comes to mind. Does it look like there are unusually long-running jobs, or unusually many jobs being queued? If the former, what jobs take a long time? If the latter, do you have any idea what jobs get enqueued a lot?

**Reply by thatcosmonaut** [reply]
When this happens, there are thousands of FeedInsertWorker jobs, I checked the logs and these are taking up to 5 seconds to complete which seems way too high. Eventually the queue usually clears up after several hours but it's quite bizarre. There's one other job I see a lot but I forget which one.

**Reply by ClearlyClaire** [reply]
Yeah, several seconds per `FeedInsertWorker` sure is a lot. I'll keep digging, but I can't really think of any reason this would happen that has not been fixed in 3.5.0rc1 :/

**Reply by offbyone** [reply]
This mirrors my experiences referenced in https://github.com/mastodon/mastodon/discussions/19797; the queue is stuffed with `FeedInsertWorker` jobs, many of which are waiting for (now) up to 30 minutes to process. 

This is running 3.5.3 on a box that doesn't show any signs of being overloaded (6 vCPUs, load average around 6 pretty consistently, no memory pressure or CPU spikes, etc...)

**Reply by ClearlyClaire** [reply]
Can you issue TTIN and to try to see where it's stuck? I really don't think `FeedInsertWorker` should do a lot of work.

https://docs.gitlab.com/ee/administration/sidekiq/sidekiq_troubleshooting.html#thread-dump

**Reply by offbyone** [reply]
https://gist.github.com/offbyone/79932a31911dc56a55c9bd0c7136d9da

**Reply by ClearlyClaire** [reply]
That doesn't seem to show any `FeedInsertWorker` currently running

**Reply by offbyone** [reply]
You're right... I was seeing the FeedInsertWorkers backlogged in the queue and thought those might be the issue. My default queue is at 30000 and growing so it's hard to get a solid picture.

**Reply by offbyone** [reply]
Using the queue API this is the count of waiting jobs in the default queue:

```
{"PushUpdateWorker"=>5092,
 "ActivityPub::ProcessingWorker"=>8258,
 "FeedInsertWorker"=>13409,
 "DistributionWorker"=>5582,
 "LocalNotificationWorker"=>802,
 "PostProcessMediaWorker"=>20,
 "MergeWorker"=>352,
 "ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper"=>44,
 "MoveWorker"=>2,
 "UnfavouriteWorker"=>2,
 "RemovalWorker"=>3,
 "PublishAnnouncementReactionWorker"=>11,
 "BootstrapTimelineWorker"=>3,
 "PushConversationWorker"=>3,
 "BlockWorker"=>1}
```

**Reply by offbyone** [reply]
Over time, these are now:

```
{"MergeWorker"=>222,
 "LocalNotificationWorker"=>577,
 "ActivityPub::ProcessingWorker"=>3662,
 "PushUpdateWorker"=>6397,
 "FeedInsertWorker"=>13362,
 "DistributionWorker"=>3777,
 "ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper"=>41,
 "PostProcessMediaWorker"=>9,
 "BootstrapTimelineWorker"=>7,
 "PushConversationWorker"=>2,
 "RemovalWorker"=>4,
 "BlockWorker"=>2,
 "UnfavouriteWorker"=>1}
```

I'm happy to provide any other troubleshooting info; I don't _think_ our instance is underscaled, although I'm not an expert. We have < 1000 users, on a VPS with 6 vCPUs and 16 GB RAM; I would hope that would be sufficient.
